src/monitors/WhammMonitor.v3

```diff
@@ -81,10 +84,90 @@ class WhammMonitor(whamm: Module) extends Monitor {
 		var bp = opcodePatterns[opcode.tag];
 
 		for (l = bp; l != null; l = l.tail) { // for every probe
-			if (Trace.whamm) Trace.OUT.put3("  match func %q+%d: %s", func.render(module.names, _), opcode_pos, opcode.mnemonic).ln();
-			def wf   = wi.functions[bp.head.2.func_index];
-			def args = bindParams(cv, wf, wi, bp.head.1, opcode, false);
-			if (args != null) Instrumentation.insertLocalProbe(module, func.func_index, opcode_pos, WhammProbe.new(wf, args));
+			if (Trace.whamm) Trace.OUT.put3("  match func %q+%d: %s", func.render(module.names, _), opcode_pos, opcode.mnemonic);
+			def wf   = wi.functions[l.head.2.func_index];
+			// check static predicate
+			if (l.head.3 != null) {
+				match (evalCall(cv, wi, opcode, l.head.3.call.target, l.head.3.call.params)) {
+					OK(v) => {
+						match (v) {
+							I32(i) => {
+								if (i == 0) {
+									if (Trace.whamm) Trace.OUT.puts(" => probe not inserted (predicate)").ln();
+									continue;
+								}
+							}
+							_ => {
+								err.at(cv.parser.decoder).WhammProbeError("whamm probe", "type error in predicate return value");
+								continue;
+							}
+						}
+					}
+					Fail => continue;
+				}
+			}
+			def args = bindParams(cv, wf, wi, l.head.1, opcode, false);
+			if (args != null) {
+				Instrumentation.insertLocalProbe(module, func.func_index, opcode_pos, WhammProbe.new(wf, args));
+				if (Trace.whamm) Trace.OUT.ln();
+			} else {
+				if (Trace.whamm) Trace.OUT.puts(" => probe not inserted (no match error)").ln();
+			}
+		}
+	}
+
+	def evalCall(cv: CodeValidator, wi: Instance, op: Opcode, target: Token, params: Array<WhammParam>) -> CallResult {
+		def exports = wi.module.exports;
+		def lookupR = FuncDecl.!(lookup(exports, target.image, Strings.equal, null));
+		if (lookupR == null) {
+			err.at(cv.parser.decoder).WhammProbeError("whamm probe", Strings.format1("static function %s not found", target.image));
+			return CallResult.Fail;
+		}
+
+		def index = lookupR.func_index;
+		def func = wi.functions[index];
+
+		def whammArgs = bindParams(cv, WasmFunction.!(func), wi, params, op, true);
+		if (whammArgs == null) return CallResult.Fail;	// propagating error, not a good idea
+		def args = if(whammArgs.length == 0, Values.NONE, Array<Value>.new(whammArgs.length));
+
+		for (i < whammArgs.length) {
+			var v: Value;
+			match (whammArgs[i]) {
+				FrameAccessor => {
+					err.at(cv.parser.decoder).WhammProbeError("whamm probe", "frame accessor used as an argument to a static function");
```

Comment:
When I was writing tests, the phrasing of this error confused me, I'd change to something like: "cannot use dynamic data in static context (FrameAccessor)"